src/main/java/nl/rkslot/nlapi_test_suite/NamelessApiTest.java

```diff
@@ -1,42 +1,126 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
 
 public class NamelessApiTest {
 
-	private static final TestStage[] TEST_STAGES = {
-//			new Announcements(),
-			new Groups(),
-			new Notifications(),
-//			new Registration(),
-			new WebsiteInfo(),
-	};
-
-	public static void main(String[] args) throws Exception {
-		if (args.length != 2) {
-			System.err.println("Usage: [api url] [enable debug]");
-			System.exit(1);
-		}
-
-		String apiUrl = args[0];
-		boolean enableDebug = Boolean.parseBoolean(args[1]);
-
-		NamelessApiBuilder builder = NamelessAPI.builder()
-				.apiUrl(apiUrl);
-		if (enableDebug) {
-			builder.withStdErrDebugLogging();
-		}
-
-		NamelessAPI api = builder.build();
-
-		for (TestStage testStage : TEST_STAGES) {
-			System.out.println("----------------- Starting test: " + testStage.getClass().getSimpleName() + " -----------------");
-			testStage.runTest(api);
-		}
-	}
+    private static final @NotNull TestStage[] TEST_STAGES = {
+            new Announcements(),
+            new Groups(),
+            new Notifications(),
+            new Registration(),
+            new WebsiteInfo(),
+            new Reports(),
+    };
+
+    public static void main(String @NotNull [] args) throws Exception {
+        if (args.length < 1) {
+            System.err.println("Usage: [api url] [enable debug]");
+            System.exit(1);
+        }
+
+        long started = Calendar.getInstance().getTimeInMillis();
+
+        final @NotNull NamelessApiBuilder builder = NamelessAPI.builder()
+                .userAgent("NamelessApiTestSuite")
+                .apiUrl(args[0]);
+
+        if (args.length == 2 && Boolean.parseBoolean(args[1])) {
+            builder.withStdErrDebugLogging();
+        }
+
+        final @NotNull NamelessAPI api = builder.build();
+
+        AtomicInteger testCount = new AtomicInteger();
+        AtomicInteger testsPassed = new AtomicInteger();
+
+        for (final @NotNull TestStage testStage : TEST_STAGES) {
+            System.out.println("----------------- Starting test class: " + testStage.getClass().getSimpleName() + " -----------------");
+
+            Arrays.stream(testStage.getClass().getDeclaredMethods()).filter(m -> {
+                return m.getName().equals("beforeRun");
+            }).findFirst().ifPresent(m -> {
+                try {
+                    if (m.getParameterCount() == 1 && m.getParameterTypes()[0].equals(NamelessAPI.class)) {
+                        m.invoke(testStage, api);
+                    } else {
+                        m.invoke(testStage);
+                    }
+                } catch (IllegalAccessException | InvocationTargetException e) {
+                    e.printStackTrace();
+                }
+            });
+
+            List<Method> tests = Arrays.stream(testStage.getClass().getDeclaredMethods()).filter(method -> {
+                return method.isAnnotationPresent(Test.class)
```

Comment:
A great way to handle multiple test methods per class, I wouldn't have thought of this!